### DISCUSSION TITLE
Why is the encoded data signed, not the raw data itself?

### DISCUSSION THREAD

**Discussion by nschloe** [keygen-sh/community#103] [General]
In license files and offline licenses, the base64-encoded data is signed. Why is that so? Couldn't we sign the raw data (JSON strings) directly?

**Comment by ezekg** [comment]
For license files, having to decode the `enc` dataset before verifying the data's signature adds a needless step when `enc` data is invalid (e.g. tampered with), and it adds one more step where somebody could unintentionally use data from inside of the encoded dataset _before_ verifying its authenticity, which is dangerous. The `enc` dataset **should always be cryptographically verified before it's decoded**, and our order of operations asserts that. And keep in mind that the `enc` property isn't always a raw JSON string — it could be an encrypted JSON string as well (i.e. a random series of bytes until successfully decrypted). In addition, we utilize a prefix during verification, and having the data be base64 encoded allows the prefixed string to be easily created — trying to concatenate a `license/` prefix to a random byte array is more challenging and error-prone, relatively, than simply concatenating 2 strings (see below for further explanation).

For signed license keys, the embedded dataset is signed after being encoded and prefixed with `key/`. This allows us to differentiate between different types of signed datum. For example, if we didn't have a prefix, then using e.g. another signature, forged into a similar structure of a license key would pass verification (e.g. `$DATA.$SIGNATURE`), even though it's not a valid license key. Encoding the data and combining it with the prefix allows us to sign the embedded data _along with_ the prefix, [protecting the scheme from forgery](https://github.com/keygen-sh/keygen-api/issues/377).

tl;dr: Both of these order of operations protect against forgery and tampering, while trying to prevent unsafe usage.
